packages/react-theming/src/create-component/createComponent.tsx

```diff
@@ -0,0 +1,140 @@
+import * as React from 'react';
+import { ProviderContext } from '../components/ThemeProvider/Provider';
+import { mergeCss } from '@uifabric/merge-styles';
+import { VariantBasedCacheKeyStrategy, ClassCache } from './ClassCache';
+
+// TODO:
+// 1. how do we know the slots for component?
+// 2. how do we tackle enum props (not just booleans)
+// 3. final decision on styles living in theme (sync with JD)
+// 4. type safety (props which are variants should be typed)
+// 5. props which are variants should not be spreaded on the root
+// 6. merging multiple variants styles should be predictable - maybe resolved...
+// 7. merging is not correct (spreading is not enough, it should be deep merge)
+// 8. How would it work for composition (Menu + MenuItem)
+// 9. support inline-style calculation based on a prop
+// 10. how to cache the styles
+
+/**
+ * Solvable:
+ * 6, 9
+ *
+ * Possible blockers:
+ * P0: 3, 8, 10
+ * P1: 4, 5, 7
+ *
+ * Solved:
+ * 1, 2
+ */
+
+const getProps = (cssMap: any, props: any, slots: any = {}) => {
+  const newProps = {
+    ...props,
+    slotProps: props.slotProps || {},
+    slots: { ...props.slots, ...slots },
+  };
+  Object.keys(cssMap).forEach(slotName => {
+    if (!newProps.slotProps[slotName]) {
+      newProps.slotProps[slotName] = {};
+    }
+    newProps.slotProps[slotName].className = `${newProps.slotProps[slotName].className ||
+      ''} ${cssMap[slotName] || ''}`;
+  });
+
+  return newProps;
+};
+
+export const getClassName = (
+  cache: ClassCache,
+  theme: any,
+  componentProps: any,
+  componentName: string,
+  cssRenderer: (args: any) => string = mergeCss,
+) => {
+  const stylesAdditions: any = {};
+  const variantNames: string[] = [];
+
+  const componentStyles =
```

Comment:
Since TypeScript 3.7 introduced optional chaining you can use it here:
const componentStyles = theme?.components?.[componentName]?.styles ? ... : ...
Ref. https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html

Reply:
Same below for variants